File: shuttle_tablet/lib/features/trips/domain/entities/trip_stop.dart
import 'package:equatable/equatable.dart';

class TripStop extends Equatable {
  final String id;
  final String tripId;
  final int sequenceOrder;
  final String locationName;
  final String? address;
  final DateTime? arrivedAt;
  final DateTime? departedAt;

  const TripStop({
    required this.id,
    required this.tripId,
    required this.sequenceOrder,
    required this.locationName,
    this.address,
    this.arrivedAt,
    this.departedAt,
  });

  @override
  List<Object?> get props =>
      [id, tripId, sequenceOrder, locationName, address, arrivedAt, departedAt];
}

File: shuttle-api/src/ShuttleApi.Domain/Trips/TripStop.cs
using ShuttleApi.Domain.Common;

namespace ShuttleApi.Domain.Trips;

public sealed class TripStop : Entity<Guid>
{
    public Guid TripId { get; private set; }
    public int SequenceOrder { get; private set; }
    public string LocationName { get; private set; } = string.Empty;
    public string? Address { get; private set; }
    public DateTime? ArrivedAt { get; private set; }
    public DateTime? DepartedAt { get; private set; }

    private TripStop() { }

    public static TripStop Create(Guid tripId, int sequenceOrder, string locationName, string? address)
    {
        return new TripStop
        {
            Id = Guid.NewGuid(),
            TripId = tripId,
            SequenceOrder = sequenceOrder,
            LocationName = locationName,
            Address = address
        };
    }

    public void SetArrived(DateTime at)
    {
        Guard.Against(ArrivedAt is not null, "Stop has already been marked as arrived.");
        ArrivedAt = at;
    }

    public void SetDeparted(DateTime at)
    {
        Guard.Against(ArrivedAt is null, "Stop must be marked as arrived before departing.");
        Guard.Against(DepartedAt is not null, "Stop has already been marked as departed.");
        DepartedAt = at;
    }

    internal void UpdateSequenceOrder(int newOrder) => SequenceOrder = newOrder;
}
